src/price_data.py

```diff
@@ -104,6 +106,69 @@ def _get_price_binance(self, base_asset: str, utc_time: datetime.datetime, quote
         average_price = total_cost / total_quantity
         return average_price
 
+    @misc.delayed
+    def _get_price_kraken(self, base_asset: str, utc_time: datetime.datetime, quote_asset: str, minutes_step: int = 10) -> float:
+        """Retrieve price from Kraken official REST API.
+
+        We select the data point closest to the desired timestamp (utc_time), but not newer than this timestamp.
+        For this we fetch one chunk of the trade history, starting 10 minutes before this timestamp.
+        We then walk through the history until the closest timestamp match is found.
+        Otherwise, we start another 10 minutes earlier and try again, etc. …
+        (Exiting with a warning and zero price after hitting the arbitrarily chosen offset limit of 120 minutes.)
+        If the 10 minutes-offset is already too large, recursively retry by reducing the offset step, down to 1 minute.
+
+        Documentation: https://www.kraken.com/features/api
+        """
+        target_timestamp = misc.to_ms_timestamp(utc_time)
+        root_url = "https://api.kraken.com/0/public/Trades"
+        pair = base_asset + quote_asset
+        pair = kraken_pair_map.get(pair, pair)
+
+        minutes_offset = 0
+        while minutes_offset < 120:
+            minutes_offset += minutes_step
+
+            since = misc.to_ms_timestamp(utc_time - datetime.timedelta(minutes=minutes_offset)) * 1000000
+            url = f"{root_url}?{pair=:}&{since=:}"
+
+            log.debug(f"Querying trades for {pair} at {utc_time} (offset={minutes_offset}m): Calling %s", url)
+            response = requests.get(url)
+            response.raise_for_status()
+            data = json.loads(response.text)
+
+            if data["error"]:
+                log.warning(f"Querying trades for {pair} at {utc_time} (offset={minutes_offset}m): "
+                            f"Could not retrieve trades: {data['error']}")
+                return 0
+
+            # Find closest timestamp match
+            data = data["result"][pair]
+            data_timestamps_ms = [int(float(d[2]) * 1000) for d in data]
+            closest_match_index = bisect.bisect_left(data_timestamps_ms, target_timestamp) - 1
+
+            # The desired timestamp is in the past; increase the offset
+            if closest_match_index == - 1:
+                continue
+
+            # The desired timestamp is in the future.
+            # Ensure at least one additional value at the end of the array to be sure our match is plausible
+            if closest_match_index >= len(data_timestamps_ms) - 2:
```

Comment:
Shouldn't this be `>`?

In the case, that to many trades were made in the 10 minutes interval. `bisect_left` returns `len(data_timestamps_ms)`. this makes `closest_match_index = len(data_timestamps_ms) - 1`.

I think we can savely change it to if `closest_match_index == len(data_timestamps_ms) - 1`. Using >= is possible here, but unnecessary I guess.

Reply:
You're right, I got confused with the idices.